### PR TITLE
[NNC] channels last propagation within NNC fusion group

### LINKED ISSUE #76438: [NHWC] Propagate the channels-last memory layout within the fusion group
Currently, we always convert the non-contiguous tensor and channels-last contiguous tensor to contiguous. But channels-last memory layout is popular and its performance is better in general. It might impact the E2E performance if we always convert the input and output tensors to contiguous.

### PR DESCRIPTION
**Pull request by EikanWang**
Decide the memory layout propagation policy and propagate it within the NNC fusion group. The memory layout propagation policy could be `Contiguous` and `Channels-last contiguous`.
 - `Contiguous`: Convert the non-contiguous including channels-last contiguous input tensors to contiguous and generate the contiguous output `Buf` for lowering function.
 - `Channels-last contiguous`: Convert the input tensors to channels-last contiguous and generate the channels-last contiguous output `Buf` for lowering function.

Currently, the rule is simple. If all the input and out tensors of the NNC fusion group are channels-last contiguous, then the propagated memory layout is `Channels-last contiguous`. Otherwise, it is always `Contiguous` which is as same as current situation. It means that this PR provides a fast path to channels-last and the optimization is conservative since its trigger conditions are strict.

### CODE REVIEW

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 1]
```diff
@@ -1482,6 +1514,97 @@ BlockPtr TensorExprKernel::bindAllInputs() {
...
+  // The graph inputs contain Tensor. Othewise, do nothing.
+  auto inputs_contain_tensor =
+      std::any_of(graph_->inputs().begin(), graph_->inputs().end(), _is_tensor);
+  auto inputs_all_channels_last = inputs_contain_tensor &&
+      std::all_of(graph_->inputs().begin(),
+                  graph_->inputs().end(),
+                  [&](const jit::Value* el) {
+                    if (el->type()->kind() != TypeKind::TensorType) {
+                      return true;
+                    }
```
Comment (jgong5) [inline torch/csrc/jit/tensorexpr/kernel.cpp:1566]: If you return `false` directly here, you don't have to check `inputs_contain_tensor` with `_is_tensor` any more, right? Basically, `inputs_all_channels_last` requires all inputs are of tensor type and in channels last format, right?

Comment (EikanWang) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1566]: > If you return `false` directly here, you don't have to check `inputs_contain_tensor` with `_is_tensor` any more, right? Basically, `inputs_all_channels_last` requires all inputs are of tensor type and in channels last format, right?

It serves the dynamic shape. The shape information is also part of graph inputs. If the logic returns false directly, then dynamic shape could not support channels-last contiguous tensor.

**Comment by EikanWang** [comment]
The "propagation" word might be not accurate. Compared to the current situation, this PR decides whether we need to convert the graph input and output tensors to channels-last format or contiguous format(current situation).

File: aten/src/ATen/core/jit_type.h [hunk 2]
```diff
@@ -804,17 +804,18 @@ struct TORCH_API TensorType : public SharedType {
       return strides;
     };
 
-    std::vector<int64_t> dim_order(sizes.size());
-    if (memory_format == MemoryFormat::ChannelsLast) {
+    auto ndims = sizes.size();
+    std::vector<int64_t> dim_order(ndims);
+    if (memory_format == MemoryFormat::ChannelsLast && ndims == 4) {
```
Comment (ZolotukhinM) [inline aten/src/ATen/core/jit_type.h:809]: Im curious, why is this change needed? I mean, it does look correct, but if something failed without it, it would be nice to add it as a regression test.

File: aten/src/ATen/core/jit_type.h [hunk 3]
```diff
@@ -804,17 +804,18 @@ struct TORCH_API TensorType : public SharedType {
...
-    } else if (memory_format == MemoryFormat::ChannelsLast3d) {
+    } else if (memory_format == MemoryFormat::ChannelsLast3d && ndims == 5) {
       dim_order = {1, 4, 3, 2, 0};
     } else {
-      auto ndims = sizes.size();
       for (size_t i = 0; i < ndims; i++) {
         dim_order[i] = ndims - i - 1; // Reverse
       }
     }
+
```
Comment (ZolotukhinM) [inline aten/src/ATen/core/jit_type.h:818]: Nit: unnecessary whitespace change.

File: test/test_tensorexpr.py [hunk 4]
```diff
@@ -1560,5 +1560,121 @@ def foo(a, b, c):
...
+            b = torch.rand(2, 3, 4, 5, dtype=torch.float32).to(memory_format=torch.channels_last)
+            c = torch.rand(2, 3, 4, 5, dtype=torch.float32)
+            # The memory layout of the imperative mode output is
+            #     (contiguous)
+            # propagation memory layout is contiguous
+            run_foo_case(foo, a, b, c)
+            # The memory layouts of the imperative mode outputs are
+            #     (contiguous, channels-last, contiguous)
+            # propagation memory layout is contiguous
+            run_foo_case(foo_multi_outputs, a, b, c)
```
Comment (ZolotukhinM) [inline test/test_tensorexpr.py:1619]: This looks like a lot of duplication. Would it be possible to factor out the common code? E.g. we could have a "config" array and then iterate over it:
```
configs = [
  ([torch.contiguous, torch.channels_last, torch.contiguous], foo), 
  ([torch.contiguous, torch.channels_last, torch.contiguous], foo_multi_outputs), 
  ([torch.contiguous, torch.channels_last, torch.contiguous], foo_multi_outputs_i_nhwc_o_nchw), 
  ...
]
```

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 5]
```diff
@@ -509,8 +509,8 @@ Tensor TensorExprKernel::computeValue(const torch::jit::Value* v) {
 
   auto outputType = findDtypeForValue(v);
   std::vector<ExprHandle> outputShape = sizesForValue(v);
-  std::vector<ExprHandle> outputStrides;
-  if (is_channels_last_contiguous && (!is_contiguous)) {
+  std::vector<ExprHandle> outputStrides = {};
+  if (memory_layout_policy_ == MemoryLayoutPolicy::kChannelsLastNdContiguous) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/kernel.cpp:513]: It looks like the loop above is not needed anymore (where we compute `is_contiguous` and `is_channels_last_contiguous`). Could you please remove it?

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 6]
```diff
@@ -963,6 +963,17 @@ std::vector<torch::jit::StrideInput>& TensorExprKernel::
...
+torch::jit::StrideInput TensorExprKernel::getSymbolicOutputStrideDesc(
+    const torch::jit::Value* value) {
+  for (size_t i : c10::irange(graph_->outputs().size())) {
+    if (value == graph_->outputs().at(i)) {
+      TORCH_INTERNAL_ASSERT(sym_stride_outputs_.count(i));
+      return sym_stride_outputs_[i];
+    }
+  }
+  TORCH_INTERNAL_ASSERT(false);
+}
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/kernel.cpp:975]: Hm, this is getting hairy, I wonder if we can make this more elegant (ie. remove duplication of how we treat inputs and outputs and avoid searching through the vector each time to just locate Value). I think this issue is beyond the scope of this PR, but if you have ideas of how it could be improved - that would be awesome :)

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 7]
```diff
@@ -1232,15 +1253,21 @@ Tensor TensorExprKernel::convertSymbolicOutputToCorrectStrides(
...
+      : at::MemoryFormat::ChannelsLast;
+  // output is contiguous with specified memory format, no work to do
+  if (buf->is_contiguous(memory_format)) {
     return Tensor(buf, nullptr);
   }
 
   TORCH_INTERNAL_ASSERT(
-      tensorOutputStrideDesc_[v->offset()] ==
-      torch::jit::StrideInput::TENSOR_CONT_CHANNELS_LAST);
+      stride_desc == torch::jit::StrideInput::TENSOR_CONT_CHANNELS_LAST);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/kernel.cpp:1270]: Nit: the name `StrideInput` is no longer accurate since it's now used for outputs too.

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 8]
```diff
@@ -1037,17 +1048,25 @@ Tensor TensorExprKernel::bindInput(const torch::jit::Value* input) {
...
   std::unordered_set<const Value*> outputs_set(outputs.begin(), outputs.end());
 
-  auto is_concrete_cont = [](const torch::jit::Value* input) {
+  auto is_concrete_cont = [](const torch::jit::Value* input,
+                             const MemoryLayoutPolicy& mem_layout_policy) {
     if (input->isCompleteTensor()) {
-      return isContiguous(input);
+      auto mem_layout = (mem_layout_policy == MemoryLayoutPolicy::kContiguous)
+          ? at::MemoryFormat::Contiguous
+          : at::MemoryFormat::ChannelsLast;
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/kernel.cpp:1056]: I wonder if we can reuse `at::MemoryFormat` here so that we don't need to introduce `MemoryLayoutPolicy`.

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 9]
```diff
@@ -1482,6 +1514,100 @@ BlockPtr TensorExprKernel::bindAllInputs() {
...
+  // The graph inputs contain Tensor. Othewise, do nothing.
+  auto inputs_contain_tensor =
+      std::any_of(graph_->inputs().begin(), graph_->inputs().end(), _is_tensor);
+  auto inputs_all_channels_last = inputs_contain_tensor &&
+      std::all_of(graph_->inputs().begin(),
+                  graph_->inputs().end(),
+                  [&](const jit::Value* el) {
+                    if (el->type()->kind() != TypeKind::TensorType) {
+                      return true;
+                    }
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/kernel.cpp:1562]: This code seems to be more complicated than it needs to be for the goal we're trying to achieve with it. Can we simply iterate over `sym_stride_inputs_` and `sym_stride_outputs_` and check if there is at least one `TENSOR_CONT_CHANNELS_LAST` in it? That's what we're trying to do here, right?

**Review by ZolotukhinM — COMMENTED** [review]
Thanks for working on this! Overall it looks good, I left some comments below! It would be good to have @eellison's and @navahgar's reviews as well, since they might be more familiar with stride-related issues.

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 5 above)
Comment (EikanWang) [reply torch/csrc/jit/tensorexpr/kernel.cpp:491]: Oh, yes. Removed.

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 6 above)
Comment (EikanWang) [reply torch/csrc/jit/tensorexpr/kernel.cpp]: I added `symbolic_strides` to `TensorExprKernel` as an attribute. The `symbolic_strides` contains the symbolic stride information of the inputs and outputs. https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/tensorexpr/kernel.cpp#L1653 So we don't need to handle inputs and output separately. But I'm not sure if I should refine more code in this PR based on this change, as I plan to make this PR more focused on channels-last and submit another PR to handle input and output in unison.

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 7 above)
Comment (EikanWang) [reply torch/csrc/jit/tensorexpr/kernel.cpp]: Besides that, the `StrideInput::TENSOR_CONT` and `StrideInput::TENSOR_CONT_CHANNELS_LAST` describe the stride information at the Tensor level, while the other enum values describe the stride information at the dimension level. https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/symbolic_shape_runtime_fusion.h#L35 It might be kind of confusing. So I'm wondering if we should always describe stride per dimension and use the utility functions to query memory layout.

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 6 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/tensorexpr/kernel.cpp]: Yeah, it can be in another PR!

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 10]
```diff
@@ -1482,17 +1477,127 @@ BlockPtr TensorExprKernel::bindAllInputs() {
...
+            (!(el->isCompleteTensor())) && (!(symbolic_strides_.count(el)));
+        if (incomplete)
+          return false;
+
+        auto preferred_mem_layout = el->isCompleteTensor()
+            ? _prefer_static_mem(el)
+            : _prefer_symbolic_mem(el, symbolic_strides_[el]);
+        return preferred_mem_layout ==
+            MemoryLayoutPolicy::kChannelsLastNdContiguous;
+      });
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/kernel.cpp:1545]: Nit: just from the code organization perspective, wouldn't it look simpler if we used a loop instead of lambdas and `std::all_of`? I think adding extra layers makes it more difficult to comprehend

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 11]
```diff
@@ -1482,17 +1477,127 @@ BlockPtr TensorExprKernel::bindAllInputs() {
...
+  auto _prefer_static_mem = [](const torch::jit::Value* val) {
+    // No shape info is present in the graph
+    TORCH_INTERNAL_ASSERT(
+        val->isCompleteTensor(),
+        buildErrorMessage(val->debugName() + " is not a complete tensor."));
+    const auto& tt = val->type()->expect<TensorType>();
+    const auto sizes = *tt->sizes().concrete_sizes();
+    const auto strides = *tt->strides().concrete_sizes();
+    return (c10::is_channels_last_strides_2d(sizes, strides))
+        ? MemoryLayoutPolicy::kChannelsLastNdContiguous
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/kernel.cpp:1505]: We're checking for strides_2d, but then set mem layour policy to Nd. Is it intentional? Will it work for non-2d cases?

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 8 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1018]: It looks like we don't really need to distinguish layouts based on the rank (if necessary, we can lookup the rank from the corresponding buffer, right?). Is there something I miss here?

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 12]
```diff
@@ -1482,17 +1477,127 @@ BlockPtr TensorExprKernel::bindAllInputs() {
...
+  GRAPH_DUMP("TensorExprKernel graph (Before graph optimization):", graph_);
+
+  // We may manipulate output pointers in graph manipulation. So we store the
+  // orignal outputs for symbolic strides information synchronization
+  std::vector<jit::Value*> _orignal_graph_outputs;
+  auto graph_outputs = graph_->outputs();
+  std::copy(
+      graph_outputs.begin(),
+      graph_outputs.end(),
+      std::back_inserter(_orignal_graph_outputs));
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/kernel.cpp:1566]: Can we just say?
```
std::vector<jit::Value*> _orignal_graph_outputs = graph_->outputs();
```

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 13]
```diff
@@ -1575,8 +1680,10 @@ void TensorExprKernel::compile() {
     if (has_symbolic_shapes_) {
       auto sizes = sizesFromSymbolicShape(tt->symbolic_sizes());
       tensorOutputSymbolicSizes_.push_back(sizes);
-      TORCH_INTERNAL_ASSERT(sym_stride_outputs_.count(i));
-      auto stride_desc = sym_stride_outputs_[i];
+      TORCH_INTERNAL_ASSERT(symbolic_strides_.count(output));
+      auto stride_desc_vec = symbolic_strides_[output];
+      TORCH_INTERNAL_ASSERT(stride_desc_vec.size() == 1);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/kernel.cpp:1685]: If the size of the desc_vec is always 1, why is it even a vector? In what scenarios can it have more than 1 element?

File: torch/csrc/jit/tensorexpr/kernel.h [hunk 14]
```diff
@@ -274,9 +292,12 @@ class TORCH_API TensorExprKernel {
   std::vector<ExprHandle> getInputStrides(
       const torch::jit::Value* input,
       const std::vector<ExprHandle>& inputTensorDims);
-  std::vector<torch::jit::StrideInput>& getSymbolicInputStrideDesc(
+  std::vector<torch::jit::StrideInput>& getSymbolicStrideDesc(
       const torch::jit::Value* value);
 
+  // Abstract graph optimizations as a single function.
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/kernel.h:298]: Nit: the comment would be more helpful if it explains why/when we need this, what we do there (just on a high level).

**Review by ZolotukhinM — APPROVED** [review]
Thanks for addressing my earlier feedback! Looks much better now! I have some more comments below, but overall I think it's good to go.

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 8 above)
Comment (EikanWang) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1018]: Suppose the graph contains 2 inputs while one is `at::MemoryFormat::ChannelsLast`(2d) and the other one is `at::MemoryFormat::ChannelsLast3d`, do you mean to use `at::MemoryFormat::ChannelsLast` to represent channels-last layout regardless the rank?

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 10 above)
Comment (EikanWang) [reply torch/csrc/jit/tensorexpr/kernel.cpp]: Replace the lambda and `std::all_of` with loop.

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 13 above)
Comment (EikanWang) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1676]: If the tensor is contiguous or channel-last contiguous, its symbolic strides always only contains 1 element - `[TENSOR_CONT]` or `[TENSOR_CONT_CHANNELS_LAST]`. But if the tensor is non-contiguous, its symbolic strides have more than 1 element. Take an example as follows, the symbolic stride is `[S_CONT, S_CONT, S_AS_ARG, S_AS_ARG, S_ONE, S_TRAN_CONT]`
```python
x = torch.randn(3, 2, 5, 4, 6, 7)
x = x.permute(1, 2, 3, 0, 5, 4)
```

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 11 above)
Comment (EikanWang) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1505]: I'm going to support 1d and 3d. The draft code could be as follows.
```c++
    return (c10::is_channels_last_strides_2d(sizes, strides)) || (c10::is_channels_last_strides_3d(sizes, strides))
        ? MemoryLayoutPolicy::kChannelsLastNdContiguous
```
